### DISCUSSION TITLE
Refreshing parent route data from child route

### DISCUSSION THREAD

**Discussion by ngunna** [remix-run/react-router#10361] [Q&A] [answered]
I have a child route that manipulates some data which happens to be displayed in the parent component. When the child route does this, I need the data in the parent component to be refreshed. In looking at the docs, I don't see anything that would assist with this particular issue, is there any kind of guidance someone can give me for a scenario like this? 

Sure, I could theoretically restructure the component/route hierarchy, but assuming that remains fixed, is there anything react-router can do for me here? The only way I currently see is to reload the entire route so as to retrigger all loaders in the route tree (not great). Seems to me you are stuck needing some cross-component messaging or more complicated state management like Redux to tell the parent to re-fetch its data otherwise. 

Thanks in advance for any help.

**Comment by huijiewei** [comment]
use react context or handle child component event

**Reply by ngunna** [reply]
Thanks for the reply, I had overlooked react context as a solution for some reason, that should work.

**Comment by brophdawg11** [comment, marked as answer]
> manipulates some data which happens to be displayed in the parent component

Are you manipulating via an `action` (`<Form method="post">`, `useSubmit`, `fetcher.submit`)?  If so, all loaders should revalidate by default and you can use [`shouldRevalidate`](https://reactrouter.com/en/main/route/should-revalidate) to opt out of revalidation on certain active loaders if they do not need to be refreshed based on the data mutation.

If you're not mutating via the built-in APIs, I would suggest trying that first.  But if you need to mutate outside of the react-router APIs, then you will need to manually revalidate via [`useRevalidator`](https://reactrouter.com/en/main/hooks/use-revalidator) and can again leverage [`shouldRevalidate`](https://reactrouter.com/en/main/route/should-revalidate) to opt-out of revalidation for certain loaders.

**Reply by ngunna** [reply]
No, I am not manipulating via an action, that would probably explain some of the confusion. This is sounding like precisely what I need though, thank you.

**Reply by brophdawg11** [reply]
Awesome!  When using `loader`'s you want to ideally perform all mutations through `action`'s as well so React Router can keep the state of your UI in sync with it's automatic revalidation logic.

**Reply by ngunna** [reply]
Yeah, totally makes sense now. In my case, I am migrating an existing app to react-router and it was already using Formik for form management/submission. It seems easiest to wrap Formik and use the useRevalidator hook in onSubmit to manually trigger the revalidation for now to achieve a similar effect as actions without having to do a bunch more migration work. However, I am sure there will be some trade-offs, so eventually working things through actions would be wise.
